adserver/analyzer/tasks.py

```diff
@@ -0,0 +1,82 @@
+"""Tasks for analyzing URLs for topics and keywords."""
+import datetime
+import logging
+
+from django.utils import timezone
+
+from ..models import Offer
+from ..models import Publisher
+from .models import AnalyzedUrl
+from .utils import get_url_analyzer_backend
+from .utils import normalize_url
+from config.celery_app import app
+
+
+log = logging.getLogger(__name__)  # noqa
+
+
+@app.task
+def analyze_url(url, publisher_slug):
+    """
+    Analyze a given URL on a publisher's site.
+
+    - Creates or updates an entry in AnalyzedUrls table
+    - Discovers keywords and topics for a URL
+    """
+    normalized_url = normalize_url(url)
+
+    publisher = Publisher.objects.filter(slug=publisher_slug).first()
+    if not publisher:
+        log.warning("Analyzing URL for nonexistent publisher: %s", publisher_slug)
+        return
+
+    backend = get_url_analyzer_backend()(url)
+    keywords = backend.analyze()  # Can be None
+
+    url_obj, created = AnalyzedUrl.objects.get_or_create(
+        url=normalized_url,
+        publisher=publisher,
+        defaults={"keywords": keywords, "last_analyzed_date": timezone.now()},
+    )
+
+    if not created:
+        url_obj.keywords = keywords
+        url_obj.last_analyzed_date = timezone.now()
+        url_obj.visits_since_last_analyzed = 0
+        url_obj.save()
+
+
+@app.task
+def daily_visited_urls(days=2):
+    """Analyze URLs that were visited recently but have never been analyzed."""
+    dt_cutoff = timezone.now() - datetime.timedelta(days=days)
+
+    # Loop over each publisher and query the offers table for distinct URLs
+    for publisher in Publisher.objects.all():
+        for obj in (
+            Offer.objects.filter(publisher=publisher, date__gt=dt_cutoff)
```

Comment:
Would it be faster do just query Offers once, and have it return the URL & the publisher?

Reply:
I'm not sure. It's worth a test.